crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move

```diff
@@ -0,0 +1,363 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[allow(unused_const)]
+module sui::bridge {
+    use std::vector;
+
+    use sui::address;
+    use sui::balance;
+    use sui::bcs;
+    use sui::bridge_committee::{Self, BridgeCommittee};
+    use sui::bridge_escrow::{Self, BridgeEscrow};
+    use sui::bridge_treasury::{Self, BridgeTreasury, token_id};
+    use sui::chain_ids;
+    use sui::coin::{Self, Coin};
+    use sui::event::emit;
+    use sui::object::{Self, UID};
+    use sui::table::{Self, Table};
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
+
+    #[test_only]
+    use sui::bridge_treasury::USDC;
+    #[test_only]
+    use sui::hex;
+    #[test_only]
+    use sui::test_scenario;
+
+    struct Bridge has key {
+        id: UID,
+        // nonce for replay protection
+        sequence_num: u64,
+        // committee pub keys
+        committee: BridgeCommittee,
+        // Escrow for storing native tokens
+        escrow: BridgeEscrow,
+        // Bridge treasury for mint/burn bridged tokens
+        treasury: BridgeTreasury,
+        pending_messages: Table<BridgeMessageKey, BridgeMessage>,
+        approved_messages: Table<BridgeMessageKey, ApprovedBridgeMessage>,
+        freezed: bool,
+        last_emergency_op_seq_num: u64,
+    }
+
+    // message types
+    const EMERGENCY_OP: u8 = 0;
+    const COMMITTEE_BLOCKLIST: u8 = 1;
+    const COMMITTEE_CHANGE: u8 = 2;
+    const TOKEN: u8 = 3;
+    const NFT: u8 = 4;
+
+    // Emergency Op types
+    const FREEZE: u8 = 0;
+    const UNFREEZE: u8 = 1;
+
+    struct BridgeMessage has copy, store, drop {
+        // 0: token , 1: object ? TBD
+        message_type: u8,
+        message_version: u8,
+        seq_num: u64,
+        source_chain: u8,
+        payload: vector<u8>
+    }
+
+    struct ApprovedBridgeMessage has store {
+        message: BridgeMessage,
+        approved_epoch: u64,
+        signatures: vector<vector<u8>>,
+    }
+
+    struct BridgeMessageKey has copy, drop, store {
+        source_chain: u8,
+        bridge_seq_num: u64
+    }
+
+    struct BridgeEvent has copy, drop {
+        message: BridgeMessage,
+        message_bytes: vector<u8>
+    }
+
+    struct TokenBridgePayload {
+        sender_address: vector<u8>,
+        target_chain: u8,
+        target_address: vector<u8>,
+        token_type: u8,
+        amount: u64
+    }
+
+    struct EmergencyOpPayload {
+        op_type: u8
+    }
+
+    const EUnexpectedMessageType: u64 = 0;
+    const EUnauthorisedClaim: u64 = 1;
+    const EMalformedMessageError: u64 = 2;
+    const EUnexpectedTokenType: u64 = 3;
+    const EUnexpectedChainID: u64 = 4;
+    const ENotSystemAddress: u64 = 5;
+    const EUnexpectedSeqNum: u64 = 6;
+
+    #[allow(unused_function)]
+    fun create(ctx: &mut TxContext) {
```

Comment:
I suppose our plan is to call this function during a epoch change transaction, similar to how Random obj is created?

Reply:
could you share a code pointer how it works?

Reply:
Random obj is created during epoch change as one of the end of epoch transactions: https://github.com/MystenLabs/sui/blob/eb67a1ea38ef998771429404eb8fd0ed96610c84/sui-execution/latest/sui-adapter/src/execution_engine.rs#L604

Reply:
ya that's the plan, I am working on it in another PR